tests.js

```diff
@@ -16,45 +16,167 @@
 
 class Tests {
   constructor(options) {
-    this.items = options.manifest.items
-        .filter((item) => !options.httpOnly || item.protocol === 'http');
-    this.individualItems = options.manifest.individualItems;
+    this.tests = options.tests;
+    this.endpoints = this.buildEndpoints();
     this.host = options.host;
+    this.httpOnly = options.httpOnly;
   }
 
-  list(after, limit) {
-    let begin; let end;
-    if (after) {
-      const afterURL = new URL(after);
-      const afterIndex = this.items.findIndex((item) => {
-        const itemURL = new URL(`${item.protocol}://${this.host}${item.pathname}`);
-        return itemURL.pathname === afterURL.pathname &&
-            itemURL.protocol === afterURL.protocol;
-      });
+  buildEndpoints() {
+    const endpoints = {};
 
-      if (afterIndex === -1) {
-        throw new Error(`${after} not found in test manifest`);
+    for (const [ident, test] of Object.entries(this.tests)) {
+      for (const scope of test.scope) {
+        let endpoint = '';
+        let httpsOnly = false;
+        switch (scope) {
+          case 'Window':
+            endpoint = '/api/interfaces';
+            break;
+          case 'Worker':
+          case 'DedicatedWorker':
+            endpoint = '/api/workerinterfaces';
+            break;
+          case 'ServiceWorker':
+            endpoint = '/api/serviceworkerinterfaces';
+            httpsOnly = true;
+            break;
+          case 'CSS':
+            endpoint = '/css/properties';
+            break;
+        }
+
+        if (endpoint) {
+          if (!(endpoint in endpoints)) {
+            endpoints[endpoint] = {
+              scope: scope,
+              httpsOnly: httpsOnly,
+              entries: []
+            };
+          }
+          if (!(ident in endpoints[endpoint].entries)) {
+            endpoints[endpoint].entries.push(ident);
+          }
+        }
       }
+    }
 
-      begin = afterIndex + 1;
-      if (limit) {
-        end = afterIndex + 1 + limit;
+    return endpoints;
+  }
+
+  listMainEndpoints(urlPrefix = '') {
+    return Object.keys(this.endpoints).map((item) => (
+      ['', `${urlPrefix}${item}`]
```

Comment:
What does the empty string mean here? I think it's a marker for "run this test by default", is that right? Being more explicit about this would be good.